src/inbox/mod_inbox_muc.erl

```diff
@@ -4,68 +4,92 @@
+      packet := Packet,
+      affiliations_map := AffsMap} = Acc) ->
+    F = fun(AffLJID, Affiliation, _) ->
+            case is_allowed_affiliation(Affiliation) of
+                true ->
+                    %% To - receiver's bare user jid (not muc jid)
+                    To = jid:make(AffLJID),
+                    %% Guess direction based on user JIDs
+                    Direction = direction(From, To),
+                    %% Host - domain of the receiver
+                    Host = To#jid.lserver,
+                    Packet2 = jlib:replace_from_to(FromRoomJid, To, Packet),
+                    update_inbox_for_user(Direction, Host, From, To, Room, Packet2);
+                false ->
+                    ok
+            end
+        end,
+    maps:fold(F, ok, AffsMap),
+    Acc.
 
+is_allowed_affiliation(outcast) -> false;
+is_allowed_affiliation(_)       -> true.
 
-update_inbox(Room, {From, FromRoomJid}, To, Packet) ->
-    Host = To#jid.server,
-    case direction(From, To) of
-        outgoing ->
-            NewPacket = jlib:replace_from_to(FromRoomJid, To, Packet),
-            handle_outgoing_message(Host, From, Room, NewPacket);
-        incoming ->
-            NewPacket = jlib:replace_from_to(FromRoomJid, To, Packet),
-            handle_incoming_message(Host, Room, To, NewPacket)
+%% Host - domain of the receiver and sender
+%% From - sender's bare user jid (not muc jid)
+%% To - receiver's bare user jid (not muc jid)
+%% Room - room's bare jid
+update_inbox_for_user(Direction, Host, From, To, Room, Packet) ->
+    %% Check that Host is a local served domain (not s2s)
+    case {is_local_host(Host), Direction} of
+        {true, outgoing} ->
+            handle_outgoing_message(Host, From, Room, Packet);
+        {true, incoming} ->
+            handle_incoming_message(Host, Room, To, Packet);
+        _ ->
+            ok
     end.
 
+%% From and To are user JIDs (not room JIDs)
 direction(From, To) ->
-    case jid:are_equal(From, To) of
+    case jid:are_bare_equal(From, To) of
         true -> outgoing;
         false -> incoming
     end.
 
-handle_outgoing_message(Host, User, Room, Packet) ->
+%% Sender and receiver is the same user
+%%
+%% Host - domain of the receiver and sender
+%% From - sender's bare user jid (not muc jid)
+%% Room - room's bare jid
+handle_outgoing_message(Host, From, Room, Packet) ->
     Markers = mod_inbox_utils:get_reset_markers(Host),
     case mod_inbox_utils:if_chat_marker_get_id(Packet, Markers) of
         undefined ->
-            mod_inbox_utils:write_to_sender_inbox(Host, User, Room, Packet);
+            mod_inbox_utils:write_to_sender_inbox(Host, From, Room, Packet);
         Id ->
-            mod_inbox_utils:reset_unread_count(User, Room, Id)
+            mod_inbox_utils:reset_unread_count(From, Room, Id)
     end.
 
+%% Host - domain of the receiver
+%% Room - room's bare jid
+%% To - receiver's bare user jid (not muc jid)
 handle_incoming_message(Host, Room, To, Packet) ->
```

Comment:
what do you think about the approach:
```
handle_incoming_message(ReceiverHost, RoomBareJid, ToBare, Packet)
```
and no comment then? It's always easy to forget to modify the comment as well

Reply:
For me it's harder to read long variables
handle_incoming_message(ReceiverHost, RoomBareJid, ToBareJid, Packet).

I can use type specs here, but heh, this PR becomes inbox refactoring....

Reply:
Sorry, when I was writing those comments, this PR had a description:
```
This PR addresses BIG CHANGES
```